Release review

mattdrayer opened the pull request:
@ormsbee @chrisndodge @cpennington -- here's the PR as described in JIRA ticket PLAT-307

chrisndodge commented on milestones/api.py:
```diff
@@ -0,0 +1,312 @@
+    milestones = data.fetch_milestones(milestone)
+    if not len(milestones):
+        return None
+    return milestones[0]
```
nitpick: if we are expecting only a single value for a unique ID (they are unique, right?), do we either want to have a data accessor which returns a scalar (as opposed to a vector), or assert that len(milestones) == 1. If they are supposed to be unique, then gettings more than one back should be a data integrity issue.

mattdrayer replied:
There's a unique database constraint on milestone.id because it's the primary key for this entity, so an integrity error would have already been thrown on the write side.  In addition, fetch_milestones() is designed to approximate the Django filter() operation which returns a set of zero or more entities.  @stephensanchez -- do you have any thoughts on how best to handle this use case?

stephensanchez replied:
Given the ID is unique, it seems logical to have a get_milestone() function exposed from the data layer.  I think maybe I'm missing something here. I thought IntegrityErrors tended to occur in our DB due to write operations, or a "get_or_create" operation. A simple "get" should be safe, right? 

Also, I wonder if it's worth having a UUID for milestones, rather than passing around the PK of the table.

mattdrayer replied:
I've added data.fetch_milestone() and wired api.get_milestone() to it.  Just a note that consumers of the Milestones app won't (or at least shouldn't) be accessing data.py directly -- I'm aok with the refactor but wanted to point out that we're only rearranging the internals here.

chrisndodge commented on milestones/api.py:
```diff
@@ -0,0 +1,312 @@
+    """
+    _validate_course_key(course_key)
+    _validate_user(user)
+    required_milestones = data.fetch_courses_milestones([course_key], 'requires', user)
```
rather than a in code string, should we have a defined constant for 'requires'. It seems like that is a system defined relationship. I see this in a few places.

mattdrayer replied:
I've added a dictionary into the data.py module to act as a proxy for these values -- let me know if it's a better approach?

chrisndodge commented on milestones/api.py:
```diff
@@ -0,0 +1,312 @@
+    """
+    Retrieves the set of milestones for list of courses
+    'relationship': optional filter on milestone relationship type (string, eg: 'requires')
+    Returns an array of dicts containing milestones
```
you might want to add a comment about how the optional 'user' parameter is used - I presume it is a filter, but it might be good to have some semantic definitions around it.

mattdrayer replied:
Done

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+            active=True
+        )
+    except internal.MilestoneRelationshipType.DoesNotExist:
+        if relationship in ['requires', 'fulfills']:
```
Might it be preferable to create these system defined relationships as part of the South migration rather than create on first reference? I believe there is a way to seed initial system data (although I haven't done so myself).

The risk here is the non-locking/transactionality around this and since we are running in a multi-worker environment, due to concurrency, multiple objects could theoretically be created.

I think this needs to be addressed.

mattdrayer replied:
I originally had a follow-on migration file (0002_seed_relationship_types.py) but it ended up being removed during one of the initial rounds of development.  I think in the beginning the thought was that the client would be determining the milestone relationship types based on various context-specific use cases.  However the approach has changed and it makes more sense to reinstate the data migration, which I'll do right now.

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+    No return currently defined for this operation
+    """
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    _delete_milestone(milestone_obj)
```
rather than having a "deep delete", I believe see you have an "is_active" flag in the schema. Might it be preferable to use that to achieve a soft delete and use the is_active flag to filter out the reads? My concern is regarding accidental author errors and loosing student achievements/fulfillments.

mattdrayer replied:
I agree that it would be nicer to go with the soft deletes (ie, set active=False instead of actually removing records).  As you've noticed all of the filtering is in place to treat inactive records as deleted on the read-side, so I think it's just a matter of making a call one way or the other -- @marcotuts, what do you think?  I'll also catch up with @ormsbee about this tomorrow.

stephensanchez replied:
Why allow deletion? I'd much rather deactivate a milestone if it has to be revoked, to preserve the data for analytics.

stephensanchez replied:
+1 to soft deletes.

mattdrayer replied:
Okay, it seems that we're all leaning toward soft deletes -- @marcotuts / @ormsbee -- unless you guys see any reason to NOT switch over to soft deletes I will go ahead and change the deletion workflows to instead set active=False.

mattdrayer replied:
FYI, soft-deletes are now available in the latest commit

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+        raise exceptions.InvalidMilestoneException()
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    if milestone_obj.id is not None:
+        return serializers.serialize_milestones(internal.Milestone.objects.filter(
```
I'll need to look at your serializers and see if you are referencing related tables in the serializers. If so, this may cause some unnecessary round_trips to the database during the serialization phase. If so, you might want to call select_related() [I think that's the name of the ORM to do the JOINs on the initial SQL query]

mattdrayer replied:
Sorry, so what is the action here?  Can you clarify?

chrisndodge replied:
Sorry, I redacted my previous question about if any related fields might be referenced (particularly in some loop). No suggested action.

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+    queryset = internal.CourseMilestone.objects.filter(
+        course_id__in=course_keys,
+        active=True
+    ).select_related('milestone')
```
+1 on the select_related as I see the is dereferenced in the serializer

mattdrayer replied:
As with your comment on line 120 -- can you clarify what action you'd like me to take here?

chrisndodge replied:
no action, +1 meant "nice to see this"

ormsbee replied:
On that note, could you please add `assertNumQueries` tests? We've got a lot of joins going on -- it'd be easy to accidentally introduce performance regressions without a test safety net.

mattdrayer replied:
Now that I am aware of assertNumQueries, I am in love with it and have sprinkled the checks all over -- it actually exposed an optimization opportunity that allowed me to chop off a number of queries from one of the workflows.

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+    # Use fetch_courses_milestones to pull the list of milestones that a user does not yet
+    # have for the specified course
+    if relationship == 'requires' and user and user.get('id', 0) > 0:
+        queryset = queryset.exclude(milestone__usermilestone__user_id=user['id'])
```
curious, why only do user filter if relationship = 'requires', might it be helpful for other types of relationships in the future?

mattdrayer replied:
We need to apply the user filtering in this case in order to see which milestones are required by a course that a user has not already collected.   At present I didn't see a need to support filtering by user for the 'fulfills' relationship, and wanted to avoid applying a filter (and doing the ensuing work) if a client attempted to filter on this combo.

Maybe it makes more sense to throw an exception for scenarios not involving 'requires' and a user filter, and future developers can modify as needed?

chrisndodge commented on milestones/serializers.py:
```diff
@@ -0,0 +1,67 @@
+        'namespace': course_milestone.milestone.namespace,
+        'description': course_milestone.milestone.description,
+        'course_id': course_milestone.course_id
+    }
```
I wonder if there is a way to query the course_milestone object (is ourse_milestone an ORM object?) to see if course_milestone.milestone has already been pre-loaded at this point, as - if it hasn't - then it could be a bit of a performance hit when calling this while iterating over a sizeable collection. This info might be useful for performance tuning.

mattdrayer replied:
Is this a situation where .select_related or prefetch_related can give us the necessary boost?

chrisndodge replied:
My main concern is for us to know when we're going to hit performance issues. Since the serializer will dereference 'milestones', it'd be good for us to know if that is going to cause a DB fetch, particularly if this is being called in a loop over a collection.

More of a discussion/curiosity than any particular feedback.

mattdrayer replied:
See the assertNumQueries additions in the latest commit/changeset

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+        milestone=milestone_obj,
+        milestone_relationship_type=relationship_type,
+        active=True,
+    )
```
might it be helpful for future use to return an object?

mattdrayer replied:
I originally had this returning an object but found I had no use cases requiring the reference so opted to leave it off -- I can go either way, let me know what you want to do.

mattdrayer replied:
Still haven't added a return here -- see below comment regarding what to return

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+            course_id=unicode(course_key),
+            content_id=unicode(content_key),
+            milestone=milestone['id'],
+            active=True,
```
if we are sticking with hard deletes, then why the active=True filter?

mattdrayer replied:
See previous note -- I've included active=True to make it easy to switch over to 'soft' deletes after we have the discussion.

mattdrayer replied:
Soft-deletes are now in place

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+        queryset = internal.Milestone.objects.filter(
+            coursecontentmilestone__milestone_relationship_type=mrt.id,
+            active=True,
+        )
```
I see we are redefining queryset here, whereas the other filters are added in a cascade fashion. Is this intention? By eye it seems like we loose the ability to filter by course_key or content_key if relationship is passed in....

mattdrayer replied:
This block actually had several problems going on due to a bit of a twisted join perspective -- I've reworked it and seems like it makes more sense now.

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+
+    # Assemble the response container
+    milestone_courses = []
+    if len(queryset):
```
nitpick: I think len(queryset) is superfluous here

mattdrayer replied:
Yep, removed.

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+        user_id=user['id'],
+        milestone=milestone_obj,
+        active=True,
+    )
```
Likewise, should we return something here?

mattdrayer replied:
As per my previous comment, I don't yet have a use case for returning an object -- can either add the return now or wait until it's needed down the road.

stephensanchez replied:
I'm a fan of returning. Same goes for the delete / deactivate, especially if we move towards deactivate.

mattdrayer replied:
So the question with "returning" in these particular workflows is what exactly to return.  For example, in create_user_milestone() we're providing the user and the milestone, and the system is creating a link between these two entities.  Should we then return the new UserMilestone relationship that was created?  Return the user?  The milestone?  Maybe we simply return True?

After we've decided on the proper pattern I can implement across the board -- I will say again that up until now I've had no reason to work with a response from these particular workflows, versus, say, the one that actually adds a new Milestone.

mattdrayer replied:
Still not returning anything in these cases

chrisndodge commented on milestones/data.py:
```diff
@@ -0,0 +1,373 @@
+        internal.UserMilestone.objects.get(
+            user_id=user['id'],
+            milestone=milestone['id'],
+            active=True,
```
likewise, why the active=True filter?

mattdrayer replied:
Per the other comments --- we just need to make a call on whether or not we want to go with the soft deletes or not.  Although either way I think it's valuable to support this filter.

mattdrayer replied:
Soft-deletes are now IN

chrisndodge commented on milestones/exceptions.py:
```diff
@@ -0,0 +1,45 @@
+    """ Exception helper """
+    raise exception(
+        'The {} you have provided is not valid: {}'.format(entity_type, entity)
+    )
```
nitpick: Seems like this is a ValidationError which has a Django defined class. You might want to consider using that class.

mattdrayer replied:
I've changed the exceptions to derive from ValidationError -- makes sense to me.

chrisndodge commented on milestones/models.py:
```diff
@@ -0,0 +1,113 @@
+    name = models.CharField(max_length=255)
+    description = models.TextField()
+    active = models.BooleanField(default=True)
+
```
Just a probing question: do any of these need an index? Do we - say - need to query all milestones by namespace? Seems like we'd also need an index on 'name'

mattdrayer replied:
I've added new db_index attributes to all of the model fields used in the various filtering scenarios

chrisndodge commented on milestones/models.py:
```diff
@@ -0,0 +1,113 @@
+    """
+    user_id = models.IntegerField(db_index=True)
+    milestone = models.ForeignKey(Milestone, db_index=True)
+    source = models.TextField(blank=True)
```
Could you explain the "source" field?

chrisndodge commented on milestones/validators.py:
```diff
@@ -0,0 +1,68 @@
+    valid_types = [
+        'requires',
+        'fulfills',
+    ]
```
likewise, seems like we should make system defined milestone types to be defined constants in code

chrisndodge commented:
@mattdrayer OK completed review pass

I think in terms of priorities:

1) Whether creating on first reference on Milestone Types is safe in a non-locking/high concurrent environment. I'd prefer this being in some migration step.

2) Verify that there might not be any "hidden" round-trips to the database because dereferenced related fields were not pre-fetched via select_related()

3) That all indexes are present based on the anticipated access patterns

4) Whether we'd rather have "soft deletes" (i.e. use the active flags). I guess that's a discussion to have with Marco

chrisndodge commented on milestones/models.py:
```diff
@@ -0,0 +1,113 @@
+    Course Entrance Exam use cases.
+    """
+    namespace = models.CharField(max_length=255)
+    name = models.CharField(max_length=255)
```
sorry, one more comment: should we have a "display_name". "name" here is more like an internal system name, correct? I could imagine that we'd at some point like to render a list of milestones for a user and having a display_name would be handy.

stephensanchez commented on milestones/tests/test_api.py:
```diff
@@ -0,0 +1,652 @@
+                'namespace': '',  # Should throw an exception
+                'description': 'Local Milestone Description'
+            })
+            self.fail('Empty Milestone Namespace: Expected InvalidMilestoneException')  # pragma: no cover
```
fyi can just use nose.tools \@raises(exceptions.InvalidMilestoneException) decorator

cpennington replied:
Or just `with self.raises(exceptions.InvalidMilestoreException):` around each of the `api.milestone` calls, to assert that they both fail. (Decorating the function only validates that at least one of them fails).

mattdrayer replied:
Thanks for the tip on using 'with' for each call -- I've changed the tests to this approach -- much cleaner

stephensanchez commented on milestones/serializers.py:
```diff
@@ -0,0 +1,67 @@
+    }
+
+
+def serialize_milestones(milestones):
```
just curious, why didn't you use DRF Serializers?

mattdrayer replied:
I ran into some confusion at the onset of the project trying to work with DRF serializers outside of the context of an HTTP request, so I went with the RYO (roll yer own) approach and haven't had time to get back around to it.  I think  we could note this as an area of tech debt and address in the next cycle, unless there's a good reason to use DRF serialization here?

stephensanchez replied:
Creating your own serializers is fine, I think I just preferred the built in serializers since we are dealing with Model Objects.  If you created one, you could simply do something in your data.py such as 
milestone_model = CourseContentMilestone.objects.get(id=milestone.id)
return CourseContentMilestoreSerializer(milestone_model).data
I guess it only comes to mind because we were talking about "user, course, location specific milestones" being a MilestoneRelationshipType between a UserMilestone and a CourseContentMilestone, in which case we may get an advantage to nesting the serializer classes to return one JSON representation of the UserMilestone and its related milestones.

mattdrayer replied:
I think we could definitely toss out the custom serializers in favor of DRF.  I'll need a hand getting it started and it will have to be added in a follow-on round of development.

ormsbee replied:
Using the DRF serializers would also let you simplify the validation (using DRF Validators). It's going to change (and hopefully simplify) a lot of the code here -- I would really like to see it done in this round if at all possible.

mattdrayer replied:
I spent a couple hours trying to transition over to using DRF serializers/validators.  Maybe I'm just thick but I wasn't able to get it to work, so I rolled back to the custom stuff.  I think this refactor should go in as a fast follower -- ie, part of our Q3 work (@marcotuts, FYI)

marcotuts replied:
@mattdrayer Thanks for calling this out. I've made a story and we can discuss prioritizing it into the next sprint as a part of Q3 follow up work. (https://openedx.atlassian.net/browse/SOL-214)

ormsbee commented on milestones/models.py:
```diff
@@ -0,0 +1,119 @@
+    namespace = models.CharField(max_length=255, db_index=True)
+    name = models.CharField(max_length=255, db_index=True)
+    description = models.TextField()
+    active = models.BooleanField(default=True, db_index=True)
```
Why this index on a boolean field?

mattdrayer replied:
Hah, I think I must have gotten carried away adding indexes per Chris' comment -- I've removed all of these boolean indexes and updated the migration file.

ormsbee commented on milestones/models.py:
```diff
@@ -0,0 +1,119 @@
+    Course Entrance Exam use cases.
+    """
+    namespace = models.CharField(max_length=255, db_index=True)
+    name = models.CharField(max_length=255, db_index=True)
```
1. Is name intended to be human readable, or some ID-like designation (e.g. `edx.passed_course.with_honors`)?
2. Shouldn't there be a unique constraint around the combo of (namespace, name)?

mattdrayer replied:
Right now the name isn't being presented to anyone anywhere, because the milestones are only being used behind the scenes to gate access to course content.  That said, I didn't want to constrain the field much more than suggesting a max length, in order to keep the door open for as many use cases as possible.

mattdrayer replied:
Added the combo constraint

ormsbee replied:
I mostly meant it by way of asking whether we wanted to add another field that was a human readable name, distinct from the id-style name and the full description.

chrisndodge replied:
I would 2nd that suggestion

mattdrayer replied:
I've added 'display_name' to the Milestone model to support the case you've described.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,334 @@
+    try:
+        return data.update_milestone(milestone)
+    except exceptions.InvalidMilestoneException:
+        raise
```
What's the point of catching this only to raise it again?

mattdrayer replied:
Yep, good catch -- this was a holdover from when data.py returned Django exceptions -- I've removed this redundancy across all of the api operations

mattdrayer commented on milestones/models.py:
```diff
@@ -0,0 +1,119 @@
+    applied to other learning objects as well, such as course content
+    (XBlocks/modules).
+    """
+    name = models.CharField(max_length=255, db_index=True)
```
Added

mattdrayer commented:
Okay, status check -- t this point I think I've addressed all of the current concrete feedback -- there are two open items requiring further discussion 1) soft deletes, and 2) missing workflow returns/responses.  I'd like to wrap up these items, plus any additional thoughts/feedback, up by EOD tomorrow (Friday) so I can focus on finishing the entrance exams feature.  @chrisndodge @ormsbee @stephensanchez, FYI

ormsbee commented on milestones/serializers.py:
```diff
@@ -0,0 +1,67 @@
+    response_data = []
+    for milestone in milestones:
+        response_data.append(serialize_milestone(milestone))
+    return response_data
```
So DRF serialization preferred, but if you do have to do this manually, you can do:

```python
    return [serialize_milestone(milestone) for milestone in milestones]
```

mattdrayer replied:
Thanks for presenting this approach -- I've gone through the code and refactored where pertinent.

ormsbee commented on milestones/serializers.py:
```diff
@@ -0,0 +1,67 @@
+        name=milestone_dict.get('name'),
+        namespace=milestone_dict.get('namespace'),
+        description=milestone_dict.get('description')
+    )
```
This would deserialize into Nones if we don't pass in the right values? Shouldn't they be blank strings, according to the model?

mattdrayer replied:
I've added default empty string values for the CharFields

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+to first consider the process for fulfilling the Milestone, and then
+consider the process for requiring the Milestone.
+
+Note the terminology difference at this layer vs. Data -- add/edit/get/remove
```
Why the difference in terminology?

mattdrayer replied:
I wanted to make it clear which layer was being worked with -- I've found it easier during development to be able to read the method and know that I'm calling the api layer or the data layer -- it also could help us in the future if we want to programmatically/automatically enforce certain access patterns (eg, all data.py modules present fetch(), and only api.py can call fetch())

ormsbee commented on milestones/data.py:
```diff
@@ -0,0 +1,390 @@
+            'description': milestone_obj.description,
+        }
+    )
+    return serializers.serialize_milestone(milestone)
```
1. Please specify in the docstring what the `milestone` var is.
2. Am I reading this right that if you create a milestone and then soft-delete it (`active=False`), and then go to create it again, it will try to create a new milestone instead of reactivating the old one?

mattdrayer replied:
Docstring updated.

Yes, that's how it currently works -- the soft delete really should be treated as deleted for all intents and purposes -- we're only doing it this way so we don't actually lose the data through a true removal.

ormsbee commented on milestones/exceptions.py:
```diff
@@ -0,0 +1,46 @@
+    """
+    CourseKey validation exception class
+    """
+    pass
```
Nitpick: Just as a note, if you have a docstring, you don't need to do a `pass`. So this would work:

```python
class InvalidCourseKeyException(ValidationError):
    """CourseKey validation exception class"""
```

mattdrayer replied:
sweet, less code :)

ormsbee commented on milestones/validators.py:
```diff
@@ -0,0 +1,67 @@
+        return False
+    if milestone.get('namespace') is None or len(milestone.get('namespace')) == 0:
+        return False
+    return True
```
`name` is also a required field, isn't it?

mattdrayer replied:
Yep, added some logic to this data validator for it

ormsbee commented on milestones/validators.py:
```diff
@@ -0,0 +1,67 @@
+        return False
+    if name not in RELATIONSHIP_TYPES.values():
+        return False
+    return True
```
Isn't this all equivalent to `return name in RELATIONSHIP_TYPES.values()`?

mattdrayer replied:
Much simpler, yes -- changed

ormsbee commented on milestones/models.py:
```diff
@@ -0,0 +1,123 @@
+    and the CourseContent, such as "requires" or "fulfills".
+    """
+    course_id = models.CharField(max_length=255, db_index=True)
+    content_id = models.CharField(max_length=255, db_index=True)
```
Do we have a way of differentiating a definition ID vs. a usage ID here? I could see reasons for both.

mattdrayer replied:
Not at the moment from this models perspective -- I suppose it would depend on the client's use case and what the client needs to store/reference as the content key given its own context and workflow(s)

ormsbee commented on milestones/models.py:
```diff
@@ -0,0 +1,123 @@
+    course_id = models.CharField(max_length=255, db_index=True)
+    milestone = models.ForeignKey(Milestone, db_index=True)
+    milestone_relationship_type = models.ForeignKey(MilestoneRelationshipType, db_index=True)
+    active = models.BooleanField(default=True)
```
Given that `CourseMilestone`, `CourseContentMilestone`, `UserMilestone`, and others in the future all share a bunch of fields (the timestamps, milestone, milestone_relationship_type, active), maybe that could be factored out into an abstract model? That would help with consistency and also give you a convenient place to put common logic like soft-deletes.

So once it's factored out, this class def would look something like:

```python
class CourseMilestone(MilestoneRelationModel):
    course_id = models.CharField(max_length=255, db_index=True)

    class Meta:
        """ Meta class for this Django model """
        unique_together = (("course_id", "milestone"),)
```

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+        'namespace': namespace,
+    }
+    _validate_milestone(milestone)
+    return data.fetch_milestones(milestone)
```
This seems odd to me.

1. We're validating the milestone here, but really we're only checking that they passed in a namespace param that's not empty or null. We're not actually validating that the milestone object is valid (nor should we -- it's just a query).
2. If we do pass a None for a namespace here, we're going to get a `InvalidMilestoneException`, but no clear message given as to why it's wrong.
3. I think it's a lot clearer if the data layer also takes `namespace` as an arg. That's not to say that the two should always be in sync. I think a common pattern later on might be to have api.py convenience functions that can query more flexible data.py functions. But in this case, I think it just makes it a little clearer what's going on.

mattdrayer replied:
Yes, I've gone around a couple times with the validators -- I think the issue stems from using them for multiple purposes -- ie, validating incoming object data on writes plus validating match criteria on reads -- I've tried to make them a little more flexible -- take a look at the latest revision and let me know if you think it's any better.  We may decide to split them into two types of validators, or only use them for writes, etc.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+            "User",
+            user,
+            exceptions.InvalidUserException
+        )
```
The indirection around how validation is done (these helper methods to validators to a shared `raise_exception` method) is somewhat confusing. Please switch to DRF serializers + validators?

mattdrayer replied:
See comment above -- tried to switch but got hung up during the refactor -- I think we should hold off on the conversion for this cycle -- unless someone else has the time/ability to jump in and make the changes, of course.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+def get_course_milestones_fulfillment_paths(course_key, user):
+    """
+    Returns a collection composed of the possible fulfillment/collection opportunites
+    """
```
Please document what it is that we're passing in (user object? username? CourseKey, unicode?) and what's being sent back.

mattdrayer replied:
Updated the docstring to include inputs and outputs -- pared-down the actual output container to only what is needed after seeing what it actually looked like.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+    # Build the set of fulfillment paths for the outstanding milestones
+    fulfillment_paths = {}
+    for milestone in required_milestones:
+        dict_key = 'milestone_{}'.format(milestone['id'])
```
Why this prefix convention for the keys?

mattdrayer replied:
I added the prefeix to make it clear what the key names actually represented (the milestone ids), at least through this review process.  If we're good with the container design it probably makes more sense to drop the prefix and go with the straight identifiers as the keys.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+            milestone,
+            get_milestone_relationship_types()['FULFILLS']
+        )
+        if len(milestone_courses):
```
Is this different from `if milestone_courses:` ?

mattdrayer replied:
Nope, changed

ormsbee commented on milestones/data.py:
```diff
@@ -0,0 +1,390 @@
+RELATIONSHIP_TYPES = {
+    'FULFILLS': "fulfills",
+    'REQUIRES': "requires",
+}
```
Why have a dict, as opposed to a list?

chrisndodge replied:
I believe these are meant to be system defined string constants.

I don't think we have a consistent best practice for constants in Open edX. Looking around the interwebs, this looks interesting:

http://code.activestate.com/recipes/65207-constants-in-python/?in=user-97991

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+    """
+    Exposes the available relationship type choices without exposing the data layer
+    """
+    return data.RELATIONSHIP_TYPES
```
If the only thing this does is return a constant, why not just make the constant a part of the API? And shouldn't relationship types be returned from the model since we have a table for it?

mattdrayer replied:
I pushed the constant down to the MilestoneRelationshipType model and plumbed the calls down through to it.  Let me know if that is what you had in mind.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+    """
+    _validate_user(user)
+    _validate_milestone(milestone)
+    return len(data.fetch_user_milestones(user, milestone))
```
The method implies it's returning a boolean val, when it's an int instead.

mattdrayer replied:
Well, technically it will evaluate to Boolean :)  Anyhoo, I changed it to return a real Boolean True/False based on the count.

ormsbee commented on milestones/api.py:
```diff
@@ -0,0 +1,325 @@
+    Returns an array of dicts
+    """
+    _validate_user(user)
+    return data.fetch_user_milestones(user)
```
This could potentially become impractically large, depending on how granular we get. Is there a use case where we'd need to do this, without specifying a namespace those milestones are in?

mattdrayer replied:
I've added mandatory filtering to this workflow -- the caller can provide an id or a namespace to filter the set down from objects.all()

ormsbee commented on milestones/data.py:
```diff
@@ -0,0 +1,390 @@
+            active=True,
+        ).delete()
+    except internal.CourseMilestone.DoesNotExist:
+        pass
```
Please add comment on why it's ok to pass here.

mattdrayer replied:
Added the comment here and the other spots where we are 'passing'

ormsbee commented on milestones/data.py:
```diff
@@ -0,0 +1,390 @@
+
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(milestone_relationship_type=mrt.id).select_related('milestone')
```
I don't think you have to specify `select_related` more than once at the top here.

mattdrayer replied:
Yep, copy/paste cruft -- removed

ormsbee commented on milestones/views.py:
```diff
@@ -0,0 +1,8 @@
+
+In this particular application, the views simply hand-off to the
+orchestration layer, which manages the application's workflows.
+"""
```
Outdated?

mattdrayer replied:
Not outdated, just nothing implemented in this first go-round -- keep the comments in place?

chrisndodge commented on milestones/migrations/0002_seed_relationship_types.py:
```diff
@@ -0,0 +1,83 @@
+        likelihood of foreign key dependencies established over time.  So for now we'll raise an
+        error and request a manual clean-up of these relationships.
+        """
+        raise RuntimeError("Cannot reverse MilestoneRelationshipType seed migration.  Please clean up all dependencies manually.")
```
I'm concerned that raising an exception might cause problems in case an emergency reverse migrations needs to happen by an Open edX operator in a production environment. I'd suggest just making backwards() a NOOP or removing the know particular milestone relationships that were created in the forwards(). Also, seems like you could also put this into the 0001_ migration file to consolidate...

ormsbee replied:
I agree with @chrisndodge -- you can make `backwards` a noop and make forwards only create the entries if they don't already exist.

mattdrayer replied:
I've updated this data migration to pass on the backwards migration -- there are no further schema changes besides the drop of all of the tables so I think it's okay to do in this case.

mattdrayer commented:
@ormsbee -- tried raising you on HipChat but perhaps you're experiencing issues like some of the other edX'ers -- I'd like to merge this review PR into master, if that's alright -- we can open up a follow-on review PR to continue discussions, if you want.  Let me know.

ormsbee commented:
@mattdrayer: I'd like to see the "milestone_1", "milestone_2" style IDs replaced with the actual milestone namespace + name (since I'm assuming that's how they're meant to be addressed externally? instead of the raw database IDs?)

Other than that, yeah, please merge. I'd still like to see changes in serialization and validation, but we can take that the next round. Thank you.

mattdrayer commented:
Ok, modified the dict keys as requested by @ormsbee to (namespace).(name).  I've squashed the commits from this review cycle and the build is green @ Travis -- going ahead with the merge.
